jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java

```diff
@@ -125,29 +125,35 @@ public class XmlConfiguration
                     Class<?> t2 = p2[i].getType();
                     if (t1 != t2)
                     {
-                        // Favour derived type over base type
-                        compare = Boolean.compare(t1.isAssignableFrom(t2), t2.isAssignableFrom(t1));
-                        if (compare == 0)
-                        {
-                            // favour primitive type over reference
-                            compare = Boolean.compare(!t1.isPrimitive(), !t2.isPrimitive());
-                            if (compare == 0)
-                                // Use name to avoid non determinant sorting
-                                compare = t1.getName().compareTo(t2.getName());
-                        }
+                        // Compare distance from Object
+                        int d1 = distanceFromObject(t1);
+                        int d2 = distanceFromObject(t2);
+                        compare = Integer.compare(d2, d1);
 
                         // break on the first different parameter (should always be true)
                         if (compare != 0)
                             break;
                     }
                 }
             }
-            compare = Math.min(1, Math.max(compare, -1));
         }
-
+        if (compare == 0)
+            compare = e1.toGenericString().compareTo(e2.toGenericString());
+        compare = Math.min(1, Math.max(compare, -1));
```

Comment:
This line is cryptic, I would add a comment that says `// Normalize to 1, 0, -1.`, since the equivalent ternary expression is cryptic too.